### PR TITLE
UITK drag and drop (ISX-1141)

### PR DESCRIPTION
**Pull request by ritamerkl**
### Description

This PR adds Drag & Drop support for the UITK-Editor. Find the ticket [here](https://jira.unity3d.com/browse/ISX-1141). 
Basically it includes reordering and moving of ActionMaps, Actions and Bindings in the Editor. 

### Changes made

added:
- reordering ActionMaps
- moving Actions to other ActionMaps
- reordering Actions
- moving Bindings and Composites to other Actions or reorder Bindings/Composites in the same Action
- moving Composite parts from one Composite to other Composites or reordering them in the same Composite

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ekcoh) [inline Packages/com.unity.inputsystem/CHANGELOG.md]: Minor but I noticed drag and drop is capitalised here but its not a named concept in the product similar to e.g. Cut, Paste, Copy where such labels exist so would make sense with just "drag and drop" or "drag-and-drop" here instead?

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:152]: I was suprised there is no destination index as argument to this function. Where do we handle a case such as copying an action inside one map, then switching to another map, selecting an existing action in the middle of that map and then pasting? I would personally expect the action to be pasted below the selected one in that case? Or maybe I am outside what was previously supported with this...

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:301]: Noticed there is a MoveComposite and a MoveBindingOrComposite but they do not seem to call into each other, how do they relate? When to use which one?

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs]: Surprised to see some kind of mouse picking is needed here (feels brittle?), isn't possible to rely on currentTarget and target for this? (I have no experience in UITK drag and drop but that's usually how this stuff works, with a drag source and a drag target). Basically I believe I do not really understand how this function operates, would it be possible to add some inline comment explaining this or feedback into this review PR?

**Review by ekcoh — COMMENTED** [review]
In general looks good and clean but have to admit I do not follow this logic completely. I was surprised around some aspects of the implementation where I have asked questions, basically not understanding why we need picking logic. I also wonder why we do not rely on DragEnter and DragExit anywhere since these kind of events are typically what are used to update drag icons and/or evaluate whether drag source and drag targets are a valid combination to e.g. discard or allow drop. Only commenting for now since I feel I need to revisit or debug to better understand implementation.

**Comment by ekcoh** [comment]
Looks like you need to run formatter on the files since formatter stage is failing which is likely the reason for additional CI issues.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (graham-huws-unity) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs]: Might be worth doing named parameters, at least for all these bools, going into ActionOrBindingData? Just to make things a bit clearer :)

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (graham-huws-unity) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: Possibly internal rather than public?

**Comment by graham-huws** [comment]
Added a couple minor things, nothing really to add beyond what Håkan mentioned - I'm always a fan of more comments explaining functions though :)

File: Packages/com.unity.inputsystem/CHANGELOG.md
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/CHANGELOG.md]: fixed in [cfbd28b](https://github.com/Unity-Technologies/InputSystem/pull/1831/commits/cfbd28b411320dc946953a23f868446fe86855f2)

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:152]: During dragging an Action the visualization does not switch to show the ActionTreeView of the Action Map drop target. So Drag&Drop of an Action is only possible into an ActionMap, not into the TreeView of that Action Map. This is consistent to the behavior of the IMGUI editor. All actions get appended to the end of the action list of the drop target.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:301]: MoveBindingOrComposite is a private Method, that is used inside of MoveComposite and MoveBinding. It's outsourced to this method to avoid redundant code.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs]: fixed in [32a3a64](https://github.com/Unity-Technologies/InputSystem/pull/1831/commits/32a3a649dfd5f67ddb511a9aa4659e88de27b84d)

**Review by Pauliusd01 — CHANGES_REQUESTED** [review]
Things I noticed so far:
- I can drop composite bindings into non composite actions (and vice versa) -> after dragging them they jump back to where they were. Expected: block symbol on mouse is shown and you can't drag them there at all
- Undo naming is not user friendly shown as "Modified Actions Maps.Array..." or "Modified x# properties"
Expected: "Moved Action/Binding x" does not need to go into specifics where it was moved, just has to be human readable of what action was done
- Unknown steps to error so far but got it when loading the ProjectWideActions test scene (ignore if cause not clear from error message alone):
```
NullReferenceException: Object reference not set to an instance of an object
UnityEngine.InputSystem.InputActionMap.Disable () (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionMap.cs:559)
UnityEngine.InputSystem.InputActionAsset.Disable () (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionAsset.cs:818)
UnityEngine.InputSystem.InputActionAsset.OnDestroy () (at ./Packages/com.unity.inputsystem/InputSystem/Actions/InputActionAsset.cs:919)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr, Boolean&) (at D:/Gitrepo/unity/Modules/IMGUI/GUIUtility.cs:219)
```
- I can move an action to its own map and then it dissapears with this error (steps to repro: Create new custom input actions asset with just one action -> move that action to its own map -> it dissapears. Does not seem to repro if the map has more than 1 action):
```
NullReferenceException: Object reference not set to an instance of an object
UnityEngine.InputSystem.Editor.CopyPasteHelper.IsComposite (UnityEditor.SerializedProperty property) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:26)
UnityEngine.InputSystem.Editor.CopyPasteHelper.PasteBindingOrComposite (UnityEditor.SerializedProperty arrayProperty, System.String json, System.Int32 index, System.String actionName, System.Boolean createCompositeParts) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:263)
UnityEngine.InputSystem.Editor.CopyPasteHelper.PasteAction (UnityEditor.SerializedProperty arrayProperty, System.String jsonToInsert, System.Int32 indexToInsert) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:253)
UnityEngine.InputSystem.Editor.CopyPasteHelper.PasteBlocks (System.String transmission, System.Int32 indexToInsert, UnityEditor.SerializedProperty arrayToInsertInto) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:217)
UnityEngine.InputSystem.Editor.CopyPasteHelper.PasteItems (System.String copyBufferString, System.Int32[] indicesToInsert, UnityEditor.SerializedProperty arrayToInsertInto) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:206)
UnityEngine.InputSystem.Editor.CopyPasteHelper.PasteData (System.String copyBufferString, System.Int32[] indicesToInsert, UnityEditor.SerializedProperty arrayToInsertInto) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:194)
UnityEngine.InputSystem.Editor.CopyPasteHelper.PasteActionsFromClipboard (UnityEngine.InputSystem.Editor.InputActionsEditorState state, System.Boolean addLast, System.Int32 mapIndex) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:179)
UnityEngine.InputSystem.Editor.CopyPasteHelper.PasteActionsOrBindingsFromClipboard (UnityEngine.InputSystem.Editor.InputActionsEditorState state, System.Boolean addLast, System.Int32 mapIndex) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/CopyPasteHelper.cs:164)
UnityEngine.InputSystem.Editor.Commands+<>c__DisplayClass12_0.<PasteActionIntoActionMap>b__0 (UnityEngine.InputSystem.Editor.InputActionsEditorState& state) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:156)
UnityEngine.InputSystem.Editor.StateContainer.Dispatch (UnityEngine.InputSystem.Editor.Command command) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/StateContainer.cs:35)
UnityEngine.InputSystem.Editor.ViewBase`1[TViewState].Dispatch (UnityEngine.InputSystem.Editor.Command command) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ViewBase.cs:78)
UnityEngine.InputSystem.Editor.ActionMapsView.OnDroppedHandler (System.Int32 mapIndex) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionMapsView.cs:75)
UnityEngine.InputSystem.Editor.DropManipulator.OnDragPerformEvent (UnityEngine.UIElements.DragPerformEvent evt) (at ./Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/DropManipulator.cs:47)
UnityEngine.UIElements.EventCallbackFunctor`1[TEventType].Invoke (UnityEngine.UIElements.EventBase evt) (at D:/Gitrepo/unity/Modules/UIElements/Core/Events/EventCallback.cs:64)
UnityEngine.UIElements.EventCallbackRegistry+DynamicCallbackList.Invoke (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, UnityEngine.UIElements.VisualElement target) (at D:/Gitrepo/unity/Modules/UIElements/Core/Events/EventCallbackRegistry.cs:228)
UnityEngine.UIElements.EventDispatchUtilities.HandleEventAcrossPropagationPath (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, UnityEngine.UIElements.VisualElement target, System.Boolean isCapturingTarget) (at D:/Gitrepo/unity/Modules/UIElements/Core/Events/EventDispatchUtilities.cs:150)
UnityEngine.UIElements.EventDispatchUtilities.DispatchToCapturingElementOrElementUnderPointer (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, System.Int32 pointerId, UnityEngine.Vector2 position) (at D:/Gitrepo/unity/Modules/UIElements/Core/Events/EventDispatchUtilities.cs:578)
UnityEngine.UIElements.MouseEventBase`1[T].Dispatch (UnityEngine.UIElements.BaseVisualElementPanel panel) (at D:/Gitrepo/unity/Modules/UIElements/Core/Events/MouseEvents.cs:248)
UnityEngine.UIElements.EventDispatcher.ProcessEvent (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel) (at D:/Gitrepo/unity/Modules/UIElements/Core/EventDispatcher.cs:336)
UnityEngine.UIElements.EventDispatcher.Dispatch (UnityEngine.UIElements.EventBase evt, UnityEngine.UIElements.BaseVisualElementPanel panel, UnityEngine.UIElements.DispatchMode dispatchMode) (at D:/Gitrepo/unity/Modules/UIElements/Core/EventDispatcher.cs:200)
UnityEngine.UIElements.BaseVisualElementPanel.SendEvent (UnityEngine.UIElements.EventBase e, UnityEngine.UIElements.DispatchMode dispatchMode) (at D:/Gitrepo/unity/Modules/UIElements/Core/Panel.cs:634)
UnityEngine.UIElements.UIElementsUtility.DoDispatch (UnityEngine.UIElements.BaseVisualElementPanel panel) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsUtility.cs:511)
UnityEngine.UIElements.UIElementsUtility.UnityEngine.UIElements.IUIElementsUtility.ProcessEvent (System.Int32 instanceID, System.IntPtr nativeEventPtr, System.Boolean& eventHandled) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsUtility.cs:230)
UnityEngine.UIElements.UIEventRegistration.ProcessEvent (System.Int32 instanceID, System.IntPtr nativeEventPtr) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsUtility.cs:76)
UnityEngine.UIElements.UIEventRegistration+<>c.<.cctor>b__1_2 (System.Int32 i, System.IntPtr ptr) (at D:/Gitrepo/unity/Modules/UIElements/Core/UIElementsUtility.cs:30)
UnityEngine.GUIUtility.ProcessEvent (System.Int32 instanceID, System.IntPtr nativeEventPtr, System.Boolean& result) (at D:/Gitrepo/unity/Modules/IMGUI/GUIUtility.cs:219)
```

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs]: good point - added in [8031068](https://github.com/Unity-Technologies/InputSystem/pull/1831/commits/8031068a041df18b4715ca4b47ff5e10066529f5)

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs]: I added a inline comment. We need to pick the item from the panel as the event is registered to the TreeView, so the target and the current target are always in this view. The issue is that we also want to detect dragging events out of this view, as we want to stop the reorderable logic in the TreeView in that case, to make dragging between the action tree view and the action map list possible. We can't choose another element to receive the events, because we want to use the reorderable logic from UITK. Picking the elements was something that was recommended by the UI-Toolkit team, find the related thread [here](https://unity.slack.com/archives/C3414V4UV/p1692199097910419?thread_ts=1692181020.348629&cid=C3414V4UV).

**Comment by ritamerkl** [comment]
> I also wonder why we do not rely on DragEnter and DragExit anywhere since these kind of events are typically what are used to update drag icons and/or evaluate whether drag source and drag targets are a valid combination to e.g. discard or allow drop.

We need to use the Drag and Drop events that are used internally in UITK for the reorderable logic. This is because we rely on the reorderable and append custom (e.g. drag between lists, restricting) implementation.

**Comment by ritamerkl** [comment]
> * I can drop composite bindings into non composite actions (and vice versa) -> after dragging them they jump back to where they were. Expected: block symbol on mouse is shown and you can't drag them there at all

I left a comment in the code where to block/discard a drag, currently it's handled with a forced reload of the treeview (that's sometimes a bit delayed what explains the jumping element). I would suggest to open a ticket for this as it's just about visuals.

> * Undo naming is not user friendly shown as "Modified Actions Maps.Array..." or "Modified x# properties"
>   Expected: "Moved Action/Binding x" does not need to go into specifics where it was moved, just has to be human readable of what action was done

I would suggest a ticket for this one as well as it's not very urgent neither.

**Comment by ritamerkl** [comment]
> * Unknown steps to error so far but got it when loading the ProjectWideActions test scene (ignore if cause not clear from error message alone):

not clear to me, looks like an problem with the asset itself.... 

> * I can move an action to its own map and then it dissapears with this error (steps to repro: Create new custom input actions asset with just one action -> move that action to its own map -> it dissapears. Does not seem to repro if the map has more than 1 action):

fixed in [2175b70](https://github.com/Unity-Technologies/InputSystem/pull/1831/commits/2175b70f05b47ee576adceb62138fa40c208288c), was actually reproducable on develop as well, seems we missed that somehow

**Comment by Pauliusd01** [comment]
https://github.com/Unity-Technologies/InputSystem/assets/54306142/8fd0c2e3-9ce2-4304-ac0d-fc97ef8b731d

Can't drag and drop actions between maps if I'm hovering over the name vs the area around it

**Comment by Pauliusd01** [comment]
https://github.com/Unity-Technologies/InputSystem/assets/54306142/fb91b602-7dbd-4327-8668-6f313e15eaec

Seems like adding new processors/interactions to an action is broken after reordering it. Probably a selection issue as deselecting and re-selecting it fixes it

**Review by Pauliusd01 — APPROVED** [review]
LGTM once bug with processors is fixed. Things checked can be found in the QA card: [ISX-1692](https://jira.unity3d.com/browse/ISX-1692). Additional bugs filed for later: [ISX-1871](https://jira.unity3d.com/browse/ISX-1871) and [ISX-1872](https://jira.unity3d.com/browse/ISX-1872)

**Comment by ritamerkl** [comment]
> Seems like adding new processors/interactions to an action is broken after reordering it. Probably a selection issue as deselecting and re-selecting it fixes it

This is definitely a selection issue. I wonder if the changes on Håkans branch (continue with -> append the selection of the moved item) would fix this easy. Maybe worth checking after a merge.

**Comment by Pauliusd01** [comment]
> > Seems like adding new processors/interactions to an action is broken after reordering it. Probably a selection issue as deselecting and re-selecting it fixes it
> 
> This is definitely a selection issue. I wonder if the changes on Håkans branch (continue with -> append the selection of the moved item) would fix this easy. Maybe worth checking after a merge.

Ok added new bug - [ISX-1873](https://jira.unity3d.com/browse/ISX-1873)

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs
Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Commands/Commands.cs:152]: I see, I think that's OK for now

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionMapsView.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionMapsView.cs:74]: I think this is likely fine but I would recommend when we do a sequence of commands like this that they are implemented as a single command instead but is the combination of the two underneath. Otherwise this would yield two Undo steps "Cut" and "Paste" which might be confusing for a Drop? Another aspect is that since dispatched, if CutActionsOrBindings() throws, the new state is not propagated and the Paste will execute on the previous state. I think I have seen this in more places so might be that we overlook all places at a later point for exception-safety and Undo convenience.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionsTreeView.cs:583]: I think its great that you have added named arguments like this, it should likely be a new default style for us to use for any function. It makes it way easier to understand when inspecting code or reviewing.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/Selectors.cs
Comment (ekcoh) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/Selectors.cs:93]: Should there be at least a Debug.Assert here on expected range of indexToInsert? Currently its used whatever it is, e.g. -21

Comment (ekcoh) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/Selectors.cs:93]: Might help catch errors later...

**Review by ekcoh — APPROVED** [review]
LGTM, added some minor comments which are more food-for-thought.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/Selectors.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/Selectors.cs:93]: fixed in [b0f2c7d](https://github.com/Unity-Technologies/InputSystem/pull/1831/commits/b0f2c7d6ca6a7da335819045ee758d6aba086252)

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionMapsView.cs
Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/ActionMapsView.cs:74]: Right, I think there your implementation of Dispatch(..,continueWith) would be a great solution, I would wait until this is landed and make use of this.

File: Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/DropManipulator.cs
Comment (graham-huws-unity) [inline Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/DropManipulator.cs:61]: Could IndexOf be used instead of the userData?

Comment (ritamerkl) [reply Packages/com.unity.inputsystem/InputSystem/Editor/UITKAssetEditor/Views/DropManipulator.cs:61]: not in this context, the userdata contains the element index, so this is saving an array operation as well

**Review by graham-huws — APPROVED** [review]

**Pull request merged**
